### ISSUE TITLE
TempDir is not accessible

### ISSUE THREAD

**Issue by davideradice** [phpmyadmin/phpmyadmin#14172]
### Steps to reproduce
Update to 4.8.0.

### Expected behaviour
Zero errors

### Actual behaviour
The $cfg['TempDir'] (./tmp/) is not accessible. phpMyAdmin is not able to cache templates and will be slow because of this.

### Server configuration
Ubuntu 16.04

Database server
Server: Localhost via UNIX socket
Server type: MariaDB
Server connection: SSL is not being used Documentation
Server version: 10.1.32-MariaDB-1~xenial - mariadb.org binary distribution
Protocol version: 10
User: root@localhost
Server charset: UTF-8 Unicode (utf8)

Web server
nginx/1.10.3
Database client version: libmysql - mysqlnd 5.0.11-dev - 20120503 - $Id: 76b08b24596e12d4553bd41fc93cccd5bac2fe7a $
PHP extension: mysqliDocumentation curlDocumentation mbstringDocumentation
PHP version: 5.6.35-1+ubuntu16.04.1+deb.sury.org+1

phpMyAdmin
Version information: 4.8.0 (up to date)
Git revision: 5ab56d0 from STABLE branch,
committed on Apr 07, 2018 at 07:02 PM by Isaac Bennetch

**Comment by ibennetch** [comment]
@davideradice Did you create the folder, and could you tell us what the permissions are right now?

**Comment by secuflag** [comment]
Into your pma installation directory:

`mkdir tmp && chmod 777 tmp`

Done.

**Comment by rogervila** [comment]
For docker users:

`mkdir /www/tmp && chmod 777 /www/tmp`

**Comment by davideradice** [comment]
@ibennetch I've not created the folder. It's the first time that I face this problem.
Now with
`sudo mkdir tmp && sudo chmod 777 tmp`
the problem is solved.
Thanks.

**Comment by davideradice** [comment]
@secuflag With "sudo" in my case. Thanks.

**Comment by pothi** [comment]
Thanks for creating this issue. As Isaac mentioned, it is due to the permission issue. I had to do the following...

`sudo mkdir tmp && sudo chown php_user tmp` in PMA installation directory.

I hope this helps someone else.

**Comment by fabius85** [comment]
This seems not be not working on Amazon EC2 instance running AMI Linux with manually installed phpMyAdmin. I tried to give the whole /usr/share/phpmyadmin folder, subfolders and files chown and setfacl to apache group and even chmod 777 to /usr/share/phpmyadmin/tmp but still get error. The only way to make it work is to set $cfg['TempDir'] = '/tmp'; in config.inc.php but i'm not sure if this is the best way to manage it

**Comment by davideradice** [comment]
@fabius85 On Linux AMI (EC2 instance), this command worked for me:
`sudo mkdir tmp && sudo chmod 777 tmp`
No need to add that row in config file.

**Comment by fabius85** [comment]
@davideradice it's probably depending on the fact you're on Ubuntu, while i'm on AMI Linux

**Comment by davideradice** [comment]
@fabius85 I've faced the problem on 2 different EC2 instances, one with ubuntu and one with linux AMI: 
> https://aws.amazon.com/amazon-linux-ami/2017.09-release-notes/

The same command solved the problem on the two instances.

**Comment by davideradice** [comment]
@fabius85 I update pma with this sh script

################################################
# Update phpMyAdmin from command line
# assuming that it's in /data/www/html/phpmyadmin/ folder
################################################
# Go to the parent folder of phpmyadmin folder
cd /data/www/html/
# Delete the old backup folder 'phpmyadmin.bak'
rm -R -d -f phpmyadmin.bak
# Create a backup of the folder 'phpmyadmin-master'
mv -f -v phpmyadmin phpmyadmin.bak
# Get the latest STABLE version of phpMyAdmin
git clone --depth=1 --branch=STABLE git://github.com/phpmyadmin/phpmyadmin.git phpmyadmin
# Copy the config file from the backup folder
cp -f -v /data/www/html/phpmyadmin.bak/config.inc.php /data/www/html/phpmyadmin/config.inc.php
# Big font for sql queries
cd /data/www/html/phpmyadmin/themes/pmahomme/css
echo "\n/* Big font for sql queries */" >> common.css.php
echo "\ncode.sql, div.sqlvalidate {font-size: x-large;}" >> common.css.php
echo "\ndiv.sqlOuter {font-size: large;}" >> common.css.php
echo "\n.CodeMirror {font-size: 150%;}" >> common.css.php
# Enter into the directory and update composer
cd /data/www/html/phpmyadmin/ && composer update
# Create tmp folder and set permission to 777
sudo mkdir tmp && sudo chmod 777 tmp
# Show the first 30 lines of the ChangeLog file
sed -n '1,30p' /data/www/html/phpmyadmin/ChangeLog

**Comment by fabius85** [comment]
@davideradice I have my own self-wrote phpMyAdmin update script which does pretty the same things yours is doing: backup old version, download new one, copy config file, update composer, create tmp dir inside phpMyAdmin folder and chmod it 777. Even though my phpMyAdmin installation folder is under `/usr/share/` instead of `/data/www/html/`
However it seems phpMyAdmin is unable to access to it's own `./tmp` dedicated folder and i ended up having to use servers' `/tmp` folder which it seems to be able to access.
Really dont get why 777 permission isnt doing it's job.

**Comment by davideradice** [comment]
@fabius85 Here are my permissions:
```
ls -l | grep phpmyadmin
drwxrwsr-x 17 ec2-user www  4096 11 apr 08.14 phpmyadmin
```

**Comment by fabius85** [comment]
@davideradice Same here:
```
ls -l | grep phpmyadmin
drwxrwsr-x.  13 ec2-user apache  4096 Apr 10 12:14 phpmyadmin
```

```
getfacl phpmyadmin
# file: phpmyadmin
# owner: ec2-user
# group: apache
# flags: -s-
user::rwx
group::rwx
other::r-x
```

**Comment by davideradice** [comment]
@fabius85 What about /tmp folder?

**Comment by fabius85** [comment]
@davideradice 

```
ls -l | grep tmp
drwxrwsrwx.  2 ec2-user apache   4096 Apr 10 12:14 tmp
```

```
getfacl tmp
# file: tmp
# owner: ec2-user
# group: apache
# flags: -s-
user::rwx
group::rwx
other::rwx
```

**Comment by fabius85** [comment]
@davideradice Now it's set, but I commented it out while doing the tests with `/usr/share/phpmyadmin/tmp`

The error shown when using `/usr/share/phpmyadmin/tmp` folder is:
`The $cfg['TempDir'] (./tmp/) is not accessible. phpMyAdmin is not able to cache templates and will be slow because of this.`

**Comment by fabius85** [comment]
Figured out it was due to SELinux. Tested it by giving `chcon -R -t httpd_sys_rw_content_t /usr/share/phpmyadmin/tmp` command, after `chmod 777 /usr/share/phpmyadmin/tmp`. No more errors reported on phpMyAdmin page. So i made the change permanent with `semanage fcontext -a -t httpd_sys_rw_content_t "/usr/share/phpmyadmin/tmp(/.*)?"` and rebooted. Error seems definitely gone.

**Comment by V1RG1L** [comment]
@pothi is right but "php_user" needs to be replaced with whatever user is the owner of the webserver process and the php process in your particular situation

**Comment by bigfella1978** [comment]
how do I do this on Mac, I am having bother cd into phpMyadmin. Can I not just go into the folder and create a new tmp file without cd'ing and do 777 chmod

**Comment by V1RG1L** [comment]
@bigfella1978 - is this a question? You should be able to sudo su on your Mac. And you should not need to chmod anything to 777. Just make the user who runs the webserver to be the owner of the tmp folder.

**Comment by bigfella1978** [comment]
in layman terms, how do you fix this

**Comment by V1RG1L** [comment]
I don't understand your question. You don't "sudo" for anything. Sudo is a program that allows you to run executables using other user privileges. Sudo su allows you to run programs as root on a Linux/UN*X type of OS. So be careful. I recommend you read this: https://www.networkworld.com/article/3236499/linux/some-tricks-for-using-sudo.html

**Comment by bigfella1978** [comment]
The question is straight forward. How do I resolve the issue in Mac OS of the temp dir. Where do I nest the tmp file and how. In layman terms

**Comment by V1RG1L** [comment]
I hope you know that underneath, the Mac OS is a UN*X type of OS. You can always open a terminal console and navigate the OS as you do a Linux system. Regarding the "layman" aspect, I don't think a "layman" should ever do these things. You need to have a minimal understanding of an UN*X system in order to do what you want to accomplish. Otherwise you need a technician.

**Comment by bigfella1978** [comment]
Let me put in a different way. Do I need to CD into a specific folder and nest a new tmp file. If so what folder and where. Im finding the whole utility a nighmare

**Comment by V1RG1L** [comment]
If you have phpMyAdmin installed on your box, that means you should have somewhere in your file system, a folder called phpMyAdmin. You need to create a folder (directory) called tmp inside this folder.

**Comment by bigfella1978** [comment]
LMAO. I have already done this Virgil. And it still does not work. I traversed into that very folder ‘phpmyadmin’ and created a new folder called ‘tmp’. Rebooted the apache servers and its still acting up. Do I need to do anything else to that folder that I created ????

**Comment by V1RG1L** [comment]
As I said, you need to identify who is the owner of the folder. Try runing ls -Filasd tm* inside the phpMyAdmin folder and post the result.

**Comment by bigfella1978** [comment]
How do you cd into that folder?

In terminal when you pull the folder in, the path explicitly says….. /Applications/XAMPP/xamppfiles/php 

 

Yet when I try and CD into it using:


cd ~/Applications/XAMPP/xamppfiles/php  


Brings up “no such file or folder"

**Comment by bigfella1978** [comment]
Let me re clarify that.


How do you cd into that folder?

In terminal when you pull the folder in, the path explicitly says….. /Applications/XAMPP/xamppfiles/phpadmin 

 

Yet when I try and CD into it using:


cd ~/Applications/XAMPP/xamppfiles/phpadmin  



No such file or directory

**Comment by V1RG1L** [comment]
Can you cd to the XAMP folder and when there run:
find . -name 'phpMyAdmin'

Be careful commands in UN* X systems are case sensitive
post the results.

**Comment by bigfella1978** [comment]
I can not CD into it as mentioned.


I pull the folder into terminal and it tells me the path is /Applications/XAMPP/xamppfiles/phpadmin 

Yet when I try cd ~/Applications/XAMPP/xamppfiles/phpadmin

I get No such file or directory


I am perplexed. What would the CD syntax be potentially???

**Comment by V1RG1L** [comment]
can you ls -Filas in the /xampfiles directory and see if phpmyadmin is not a symlink?

I'm using MAMP so I'm running a bit blind here.

**Comment by bigfella1978** [comment]
Here see screenshot.

**Comment by V1RG1L** [comment]
?

**Comment by PurpleCapedCrusader** [comment]
Hey @bigfella1978 and @V1RG1L 

I just ran into the same issue and I think I can help you get past Changing the Directory to the phpmyadmin folder. 

1. In the terminal type "**ls**" to list the folders of the directory you're currently in.
2. If you see folders named Applications, Desktop, Documents, etc... then you're in your user folder and you need to back out with "**cd /**". This will take you to the folder with right Applications folder; the one for all users, not just your user.
3. From here you should be able to get to the phpmyadmin folder with "**cd Applications/XAMPP/xamppfiles/phpmyadmin/**" 

I can't fix the tmp folder issue yet, but hopefully that solves your issue with finding the phpmyadmin folder.

COOL TIP:
You may already know about it, but there's a Terminal shortcut to quickly type the folder names by typing the first 2 or 3 letters and then pressing "**tab**" to autofill the rest. You only have to type enough letters so there is only 1 possibility it could be. Example: if you want to get to an "Applications" folder and there's also an "App" folder, you'll need to type "Appl" then press "tab". If there is only 1 folder beginning with "A" then you can just type "A" and press tab.

**Comment by PurpleCapedCrusader** [comment]
I did end up getting it working. I created the tmp folder in the phpmyadmin folder and changed the owner to root... that didn't work... until I changed the Sharing & Permissions to give Read & Write to everyone it started working.

**Comment by bigfella1978** [comment]
There is a lot of verbose answers on how to resolve this when in fact it is simple.

1. Open up the config.inc.php file in the text editor. Note: config.inc.php is nested within phpMyAdmin folder

2. Add the line of code `$cfg['TempDir'] = '/tmp';

3. Restart the all servers in phpMyAdmin  and the error goes.

**Comment by V1RG1L** [comment]
I don't think you should add it in the config.inc.php file.

Instructions from the config.txt file inside the phpmyadmin install folder:

".. _web-dirs:

Web server upload/save/import directories
-----------------------------------------

If PHP is running in safe mode, all directories must be owned by the same user
as the owner of the phpMyAdmin scripts.

If the directory where phpMyAdmin is installed is subject to an
``open_basedir`` restriction, you need to create a temporary directory in some
directory accessible by the PHP interpreter."

".. config:option:: $cfg['TempDir']

    :type: string
    :default: ``''``

    The name of the directory where temporary files can be stored.

    This is needed for importing ESRI Shapefiles, see :ref:`faq6_30` and to
    work around limitations of ``open_basedir`` for uploaded files, see
    :ref:`faq1_11`.

    This directory should have as strict permissions as possible as the only
    user required to access this directory is the one who runs the webserver.
    If you have root privileges, simply make this user owner of this directory
    and make it accessible only by it:

    .. code-block:: sh

        chown www-data:www-data tmp
        chmod 700 tmp

    If you cannot change owner of the directory, you can achieve a similar
    setup using :term:`ACL`:

    .. code-block:: sh

        chmod 700 tmp
        setfacl -m "g:www-data:rwx" tmp
        setfacl -d -m "g:www-data:rwx" tmp

    If neither of above works for you, you can still make the directory
    :command:`chmod 777`, but it might impose risk of other users on system
    reading and writing data in this directory.

    .. warning::

        Please see top of this chapter (:ref:`web-dirs`) for instructions how
        to setup this directory and how to make its usage secure."

**Comment by hikmetkutuk** [comment]
For archlinux;
Edit
/etc/webapps/phpmyadmin/config.inc.php to add the line:

 $cfg['TempDir'] = '/tmp/phpmyadmin'

**Comment by ThermoMan** [comment]
chmod 777 is generally a bad idea.
You should give the minimimum permissions necessary.

I'm on Ubuntu 18.04.1
phpMyAdmin 4.8.3

```
cd /usr/share/phpmyadmin
sudo mkdir tmp
sudo chgrp www-data tmp
sudo chmod 550 tmp

ls -ld tmp
dr-xr-x--- 3 root www-data 4096 Dec 10 00:05 tmp

```
